src/Disks/ObjectStorages/Encrypted/EncryptedObjectStorage.cpp

```diff
@@ -0,0 +1,199 @@
+#include "EncryptedObjectStorage.h"
+
+#if USE_SSL
+
+#    include <filesystem>
+#    include <Disks/IO/CachedOnDiskWriteBufferFromFile.h>
+#    include <Disks/ObjectStorages/DiskObjectStorageCommon.h>
+#    include <IO/BoundedReadBuffer.h>
+#    include <IO/ReadBufferFromFileBase.h>
+#    include <IO/WriteBufferFromEncryptedFile.h>
+#    include <Interpreters/Context.h>
+#    include <Common/CurrentThread.h>
+#    include <Common/logger_useful.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int DATA_ENCRYPTION_ERROR;
+}
+
+String EncryptedObjectStorageSettings::findKeyByFingerprint(UInt128 key_fingerprint, const String & path_for_logs) const
+{
+    auto it = all_keys.find(key_fingerprint);
+    if (it == all_keys.end())
+    {
+        throw Exception(
+            ErrorCodes::DATA_ENCRYPTION_ERROR, "Not found an encryption key required to decipher file {}", quoteString(path_for_logs));
+    }
+    return it->second;
+}
+
+EncryptedObjectStorage::EncryptedObjectStorage(
+    ObjectStoragePtr object_storage_, EncryptedObjectStorageSettingsPtr enc_settings_, const std::string & enc_config_name_)
+    : object_storage(object_storage_), enc_settings(enc_settings_), enc_config_name(enc_config_name_), log(&Poco::Logger::get(getName()))
+{
+}
+
+DataSourceDescription EncryptedObjectStorage::getDataSourceDescription() const
+{
+    auto wrapped_object_storage_data_source = object_storage->getDataSourceDescription();
+    wrapped_object_storage_data_source.is_encrypted = true;
+    return wrapped_object_storage_data_source;
+}
+
+std::string EncryptedObjectStorage::generateBlobNameForPath(const std::string & path)
+{
+    return object_storage->generateBlobNameForPath(path);
+}
+
+ReadSettings EncryptedObjectStorage::patchSettings(const ReadSettings & read_settings) const
+{
+    ReadSettings modified_settings{read_settings};
+    modified_settings.encryption_settings = enc_settings;
+    return object_storage->patchSettings(modified_settings);
+}
+
+void EncryptedObjectStorage::startup()
+{
+    object_storage->startup();
+}
+
+bool EncryptedObjectStorage::exists(const StoredObject & object) const
+{
+    return object_storage->exists(object);
+}
+
+std::unique_ptr<ReadBufferFromFileBase> EncryptedObjectStorage::readObjects( /// NOLINT
+    const StoredObjects & objects,
+    const ReadSettings & read_settings,
+    std::optional<size_t> read_hint,
+    std::optional<size_t> file_size) const
+{
+    return object_storage->readObjects(objects, patchSettings(read_settings), read_hint, file_size);
+}
+
+std::unique_ptr<ReadBufferFromFileBase> EncryptedObjectStorage::readObject( /// NOLINT
+    const StoredObject & object,
+    const ReadSettings & read_settings,
+    std::optional<size_t> read_hint,
+    std::optional<size_t> file_size) const
+{
+    return readObjects({object}, read_settings, read_hint, file_size);
```

Comment:
As far as I see `patchSettings` is not called by `DiskObjectStorage` - only but `IObjectStorage's`, so if you create a `DiskObjectStorage` with `EncryptedObjectStorage`, but without `CachedObjectStorage`, then nobody will do `patchSettings` for `EncryptedObjectStorage`, so decryption will not work. 
Let's make `patchSettings` always called by `readObject`. Also let's remove this line https://github.com/ClickHouse/ClickHouse/blob/master/src/Disks/ObjectStorages/Cached/CachedObjectStorage.cpp#L58 so that each object storage implementation applies only its own settings.

Reply:
ahh sorry, I misread the code. All correct.